File: web-app/src/utils/LruCache.ts
export class LruCache<K, V> {
  private capacity: number;
  private map: Map<K, V> = new Map();
  private expiry: Map<K, number> = new Map();

  constructor(capacity = 128) {
    this.capacity = capacity;
  }

  get(key: K): V | undefined {
    const value = this.map.get(key);
    const exp = this.expiry.get(key);
    if (value !== undefined && exp && exp > Date.now()) {
      this.map.delete(key);
      this.map.set(key, value); // refresh order
      return value;
    }
    if (value !== undefined) {
      this.delete(key);
    }
    return undefined;
  }

  put(key: K, value: V, ttlMs: number) {
    if (this.map.size >= this.capacity) {
      const oldest = this.map.keys().next().value;
      if (oldest !== undefined) {
        this.delete(oldest);
      }
    }
    this.map.set(key, value);
    this.expiry.set(key, Date.now() + ttlMs);
  }

  delete(key: K) {
    this.map.delete(key);
    this.expiry.delete(key);
  }
}

File: mobile-app/packages/services/lib/src/lru_cache.dart
class LruCache<K, V> {
  final int capacity;
  final _map = <K, V>{};
  final _expiry = <K, DateTime>{};

  LruCache([this.capacity = 128]);

  V? get(K key) {
    final value = _map.remove(key);
    if (value != null) {
      final exp = _expiry[key];
      if (exp != null && exp.isAfter(DateTime.now())) {
        _map[key] = value;
        return value;
      } else {
        _expiry.remove(key);
      }
    }
    return null;
  }

  void put(K key, V value, Duration ttl) {
    if (_map.length >= capacity) {
      final firstKey = _map.keys.first;
      _map.remove(firstKey);
      _expiry.remove(firstKey);
    }
    _map[key] = value;
    _expiry[key] = DateTime.now().add(ttl);
  }

  void delete(K key) {
    _map.remove(key);
    _expiry.remove(key);
  }
}
